Avoid errors due to bad request_id during signup

zachmargolis opened the pull request:
**Why**: Caused exceptions in production

---

So this will silently eat errors and avoid associating an SP with the session, but continue to confirm the user. I am not convinced this is the right approach, because we may not have enough information to forward the user to the right place after they sign up. I think silently eating the error like this is probably kicking the can down the line.

Another option would be to show an error asking users to check the URL? I am not sure what experience we want here, or what the reasons for having malformed `request_id` params are

dnesting commented on app/services/store_sp_metadata_in_session.rb:
```diff
@@ -5,6 +5,8 @@ def initialize(session:, request_id:)
   end
 
   def call
+    return unless sp_request
```
This PR gives me a little heartburn.  If this is a genuine problem, we should at least *log* something about it so we know when it happens (and how often), and possibly enough information that we can identify what user is in this state.  Returning this error to the user with a 500 might be reasonable until we figure out what's causing it.  If the user is providing a broken/invalid request ID, maybe the right thing is just to generate another error with a 200 or 4xx response code.  The only thing I'd prefer we avoid is an uncaught exception that then pages us through New Relic.

zachmargolis replied:
Yeah, see comment above. I'm glad that somebody else thought silently eating the error was a bad idea too.

I think we can catch this error earlier on by validating the request ID when we validate the confirmation token

monfresh replied:
Yes, we should validate the request ID. Note that this should rarely happen since the link in the confirmation email puts the `request_id` before `confirmation_token`, and if the user only partially copies over the URL, it will fail because we validate the confirmation token. This particular error in `int` was caused by the same user who was not following the regular flow.

monfresh replied:
Perhaps we can rename `EmailConfirmationTokenValidator` to `EmailConfirmationValidator` and add in a validation for the request ID.

monfresh replied:
I take that back. That would not solve the issue where a user deliberately changes the request ID in the URL for the password creation page. It would need to be a new `RequestIdValidator` class called in the `edit` action of `Users::PasswordsController`

zachmargolis commented:
Ok @monfresh @dnesting I reverted the last fix and added a commit that checks the `_request_id` in the email confirmations controller so we can try to catch the error upstream here.

We could be extra defensive and add a similar check back in to the `SignUp::PasswordsController`, but I'm not sure there's as clear of a course of action there

monfresh commented on app/controllers/sign_up/email_confirmations_controller.rb:
```diff
@@ -18,6 +20,15 @@ def create
+      return if result.success?
+      flash[:error] = t('errors.messages.request_id_invalid')
+      redirect_to sign_up_email_resend_url
+    end
```
Looks like you forgot to check in the RequestIdValidator class?

zachmargolis replied:
whooooops, fixed in b3ef60bb6223e7f72768cef762e98d1dc7d1ed6b

monfresh commented on app/services/request_id_validator.rb:
```diff
@@ -0,0 +1,37 @@
+
+  def extra_attributes
+    {
+      request_id_present: request_id.present?
```
How about capturing the actual request_id to help us troubleshoot? I think we should also capture it when it first gets set. That way, we can compare and see if it's being modified on the password screen.

zachmargolis replied:
Done in 378424b83c5d3a27834066c5b542f7b16d5063ed

monfresh commented on spec/factories/service_providers.rb:
```diff
@@ -8,4 +8,11 @@
+  factory :service_provider_request do
+    uuid { SecureRandom.uuid }
+    issuer { 'test issuer name' }
+    loa { 1 }
```
To avoid confusion and for better accuracy, can we make this a string column and maybe use a default value of `Saml::Idp::Constants::LOA1_AUTHN_CONTEXT_CLASSREF` or `http://idmanagement.gov/ns/assurance/loa/1`, which is what the column would contain.

zachmargolis replied:
Sure, will do

monfresh commented on app/controllers/sign_up/email_confirmations_controller.rb:
```diff
@@ -18,6 +20,15 @@ def create
+
+      return if result.success?
+      flash[:error] = t('errors.messages.request_id_invalid')
+      redirect_to sign_up_email_resend_url
```
I'm not sure this is the right thing to do. We want them to double check that they clicked the link in the email, or if they copied and pasted it, that they did so correctly. By redirecting them to resend the email confirmation, we lose the request_id, and they won't be redirected back to the SP when they are done.

I think we should leave them on the password creation page with the flash message that explains what the error is.

monfresh replied:
Clarification: they won't be redirected when they click on the email confirmation in a different browser than the one they started with, which will be the case for everyone using the CBP Jobs app.

zachmargolis replied:
I can remove the redirect. How would you phrase the error message?

monfresh replied:
How about "Oops, something went wrong. Please try clicking the link in your confirmation email again."

monfresh commented on app/services/request_id_validator.rb:
```diff
@@ -0,0 +1,37 @@
+
+  def extra_attributes
+    {
+      request_id: request_id,
```
I thought of a better idea. How about tracking the request_id automatically everywhere by changing `request_attributes` in `analytics.rb` to this:
```ruby
def request_attributes
  {
    user_ip: request.remote_ip,
    user_agent: request.user_agent,
    host: request.host,
    request_id: request.parameters.fetch(:request_id, nil),
  }
end
```

zachmargolis replied:
A very limited number of pages use the `request_id` in the URL, do you think it's worth it?

monfresh replied:
Not anymore. This whole PR needs to be scrapped. See my latest comment. I figured out the root cause. It's a bug in our code.

monfresh replied:
I can work on it if you have better things to do.

zachmargolis replied:
So shall it be scrapped!

monfresh commented:
I figured out the issue. It's a bug in our code. Here are the steps to reproduce:

Browser 1: Go to SP, go to Login.gov, choose to create account, enter email

Browser 2: Click on confirmation link, finish setting up account, get redirected back to SP. stay signed in

Browser 1: Go to SP, click on Login.gov, choose to create account, enter email

Browser 2: Note that the `request_id` in this second confirmation email is the same as the one in the previous email. Click the confirmation link, then try to set your password -> BOOM.

This is because we delete the `ServiceProviderRequest` from the DB after the user is redirected to the SP, and then if we see that the session already has a `request_id`, we don't store it again in the DB and don't generate a new one, so the next time the app looks it up, it doesn't find a match in the DB and it causes the exception
